File: 5-Problem-A-Day/286. Walls and Gates.java
class Solution {
    int[][] dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public void wallsAndGates(int[][] rooms) {
        int m = rooms.length, n = m == 0 ? 0 : rooms[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rooms[i][j] == 0)
                    dfs(rooms, m, n, i, j, 0);
            }
        }
    }

    private void dfs(int[][] rooms, int m, int n, int i, int j, int dis) {
        if (i < 0 || i >= m || j < 0 || j >= n || rooms[i][j] == -1
                || (rooms[i][j] <= dis && dis != 0))
            return;
        rooms[i][j] = dis;
        for (int[] d : dir) {
            int x = i + d[0], y = j + d[1];
            dfs(rooms, m, n, x, y, dis + 1);
        }
    }
}

File: 2021/286. Walls and Gates.cpp
class Solution {
public:
    int directions[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    void wallsAndGates(vector<vector<int>>& rooms) {
        int m = rooms.size(), n = rooms[0].size();        
        queue<pair<int, int>> steps;        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rooms[i][j] == 0) {
                    steps.push({i, j});
                }
            }            
        }
        int level = -1;        
        while (!steps.empty()) {
            level++;
            int size = steps.size();
            for (int i = 0; i < size; i++) {
                int row = steps.front().first, col = steps.front().second;
                steps.pop();
                for (int d = 0; d < 4; d++) {
                    int newRow = row + directions[d][0], newCol = col + directions[d][1];
                    if (newRow < 0 || newRow >= m || newCol < 0 || newCol >= n || rooms[newRow][newCol] != INT_MAX) {
                        continue;
                    }
                    rooms[newRow][newCol] = level + 1;
                    steps.push({newRow, newCol});
                }
            }
        }        
    }
};